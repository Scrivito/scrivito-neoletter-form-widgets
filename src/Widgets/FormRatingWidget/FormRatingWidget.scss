.scrivito-neoletter-form-widgets .form-rating {
  .rating-title {
    display: flex;
    align-items: baseline;
  }
}
