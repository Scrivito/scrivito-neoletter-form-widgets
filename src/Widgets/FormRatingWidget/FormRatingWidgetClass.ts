import * as Scrivito from "scrivito";

export const FormRatingWidget = Scrivito.provideWidgetClass(
  "FormRatingWidget",
  {
    attributes: {
      title: "string",
      colorType: [
        "enum",
        {
          values: ["default", "primary", "secondary", "custom"]
        }
      ],
      size: [
        "enum",
        { values: ["bs-icon-1x", "bs-icon-default", "bs-icon-2x", "bs-icon-3x", "bs-icon-4x", "bs-icon-5x"] }
      ],
      icon: "string",
      customColor: "string",
      hoverEffect: "boolean",
      helpText: "html",
      customFieldName: "string"
    }
  }
);
