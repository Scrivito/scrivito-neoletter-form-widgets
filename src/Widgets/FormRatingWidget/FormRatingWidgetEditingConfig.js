import * as Scrivito from "scrivito";
import { customFieldNameValidation } from "../FormContainerWidget/utils/validations/customFieldNameValidation";
import { insideFormContainerValidation } from "../FormContainerWidget/utils/validations/insideFormContainerValidation";
import formContainerWidgetIcon from "../../assets/images/form_widget_rating.svg";

Scrivito.provideEditingConfig("FormRatingWidget", {
  title: "Form Rating",
  thumbnail: formContainerWidgetIcon,
  attributes: {
    title: { title: "Label" },
    customFieldName: { title: "Field name" },
    helpText: { title: "Help text" },
  },
  properties: ["title", "customFieldName", "helpText"],
  initialContent: {
    title: "Please leave your rating",
    customFieldName: "custom_",
  },
  validations: [insideFormContainerValidation, customFieldNameValidation],
});
