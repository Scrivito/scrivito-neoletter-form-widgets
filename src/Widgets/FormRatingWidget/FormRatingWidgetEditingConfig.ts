import * as Scrivito from "scrivito";
import { customFieldNameValidation } from "../FormStepContainerWidget/utils/validations/customFieldNameValidation";
import { insideFormContainerValidation } from "../FormStepContainerWidget/utils/validations/insideFormContainerValidation";
import formContainerWidgetIcon from "../../assets/images/form_widget_rating.svg";
import { IconEditorTab } from "./IconEditorTab/IconEditorTab";

Scrivito.provideEditingConfig("FormRatingWidget", {
  title: "Neoletter Form Rating",
  thumbnail: formContainerWidgetIcon,
  attributes: {
    title: { title: "Label" },
    customFieldName: { title: "Field name" },
    helpText: { title: "Help text" },
    hoverEffect: {
      title: "Hover effect",
      description: "Previews icon selection on hover."
    },
    size: {
      title: "Size",
      description: "Default: 133%",
      values: [
        { value: "bi-1x", title: "100% (Normal)" },
        { value: "bi-default", title: "133%" },
        { value: "bi-2x", title: "200%" },
        { value: "bi-3x", title: "300%" },
        { value: "bi-4x", title: "400%" },
        { value: "bi-5x", title: "500% (Biggest)" }
      ]
    },
    colorType: {
      title: "Color",
      description: "Choose the color for selected icons.",
      values: [
        { value: "default", title: "Default" },
        { value: "primary", title: "Primary" },
        { value: "secondary", title: "Secondary" },
        { value: "custom", title: "Custom" }
      ]
    },
    customColor: { title: "Custom Hex Color" }
  },
  properties: ["title", "customFieldName", "helpText"],
  propertiesGroups: (widget: Scrivito.Widget) => {
    return [
      {
        title: "Stylings",
        key: "form-icon-styling",
        properties: getStylingProps(widget)
      },
      {
        title: "Icon",
        properties: [],
        component: IconEditorTab,
        key: "form-icon-group"
      }
    ];
  },
  initialContent: {
    title: "Please leave your rating",
    customFieldName: "custom_",
    colorType: "default",
    customColor: "#ffd700",
    hoverEffect: false,
    icon: "bi-star-fill",
    size: "bi-default"
  },
  validations: [
    insideFormContainerValidation,
    customFieldNameValidation,
    [
      "customColor",
      (customColor: string, { widget }: { widget: Scrivito.Widget }) => {
        if (widget.get("colorType") != "custom") return;
        if (!isHexColor(customColor)) {
          return "Specify a valid hex color.";
        }
      }
    ]
  ]
});

function isHexColor(color: string): boolean {
  return /^#([0-9A-F]{3}){1,2}$/i.test(color);
}

function getStylingProps(widget: Scrivito.Widget): string[] {
  const props = ["hoverEffect", "size", "colorType"];
  if (widget.get("colorType") == "custom") {
    props.push("customColor");
  }
  return props;
}
