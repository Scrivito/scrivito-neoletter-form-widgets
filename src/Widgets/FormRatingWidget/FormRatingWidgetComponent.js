import * as React from "react";
import * as Scrivito from "scrivito";
import { Star } from "../FormContainerWidget/components/StarComponent";
import { getFieldName } from "../FormContainerWidget/utils/getFieldName";
import { HelpText } from "../FormContainerWidget/components/HelpTextComponent";
import "./FormRatingWidget.scss";

Scrivito.provideComponent("FormRatingWidget", ({ widget }) => {
  const [selectedStars, setSelectedStars] = React.useState(0);

  return (
    <div className="form-rating mb-3">
      <div className="rating-title">
        <span className="text-super">{widget.get("title")}</span>
        {widget.get("helpText") && <HelpText widget={widget} />}
      </div>
      {createArray(5).map((n, i) => (
        <Star
          key={i}
          color={"gold"}
          selected={selectedStars > i}
          onSelect={() => setSelectedStars(i + 1)}
        />
      ))}
      <input
        type="hidden"
        name={getFieldName(widget)}
        value={selectedStars == 0 ? "" : selectedStars}
      ></input>
    </div>
  );
});

const createArray = (length) => [...Array(length)];
