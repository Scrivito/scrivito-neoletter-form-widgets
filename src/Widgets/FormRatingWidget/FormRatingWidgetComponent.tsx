import * as React from "react";
import * as Scrivito from "scrivito";
import { RatingIcon } from "../FormStepContainerWidget/components/RatingIconComponent";
import { getFieldName } from "../FormStepContainerWidget/utils/getFieldName";
import { HelpText } from "../FormStepContainerWidget/components/HelpTextComponent";
import { FormRatingWidget } from "./FormRatingWidgetClass";
import { getIconColor } from "../FormStepContainerWidget/utils/getIconColor";
import "./FormRatingWidget.scss";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
Scrivito.provideComponent(FormRatingWidget, ({ widget, onInputChange }: any) => {
  const [selectedIcons, setSelectedIcons] = React.useState(0);
  const [hoveredIcons, setHoveredIcons] = React.useState(0);
  const hoverEffect = widget.get("hoverEffect");
  const color = getIconColor(widget);

  const handleSelect = (rating: number) => {
    setSelectedIcons(rating);
    onInputChange && onInputChange(getFieldName(widget), rating == 0 ? "" : rating.toString());

  }
  return (
    <div className="form-rating mb-3">
      {widget.get("title") && <div className="rating-title">
        <Scrivito.ContentTag
          attribute="title"
          content={widget}
          tag="span"
        />
        {widget.get("helpText") && <HelpText widget={widget} />}
      </div>
      }
      {createArray(5).map((n, i) => (
        <RatingIcon
          key={i}
          icon={widget.get("icon") || "bi-star-fill"}
          color={getColor(
            color,
            selectedIcons > i,
            hoverEffect && hoveredIcons > i
          )}
          size={widget.get("size") || "bs-icon-default"}
          onSelect={() => handleSelect(i + 1)}
          onHover={() => hoverEffect && setHoveredIcons(i + 1)}
          onHoverOut={() => hoverEffect && setHoveredIcons(0)}
        />
      ))}
      <input
        type="hidden"
        className="show-in-review"
        name={getFieldName(widget)}
        value={selectedIcons == 0 ? "" : selectedIcons}></input>
    </div>
  );
});

const createArray = (length: number) => [...Array(length)];

const getColor = (
  color: string,
  selected: boolean,
  hoveredOver: boolean
): string => {
  return selected || hoveredOver ? color : "";
};
