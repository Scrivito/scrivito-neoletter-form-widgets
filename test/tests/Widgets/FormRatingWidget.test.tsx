import { screen, fireEvent } from "@testing-library/react";
import * as Scrivito from "scrivito";
import { FormRatingWidget } from "../../../src/Widgets/FormRatingWidget/FormRatingWidgetClass";
import "../../../src/Widgets/FormRatingWidget/FormRatingWidgetComponent";
import PageRenderer from "../../helpers/pageRenderer";

Scrivito.configure({ tenant: "inMemory" });

const pageRenderer = new PageRenderer();

const widgetProps = {
  title: "Test Rating Title",
  helpText: "Sample help text",
  icon: "bi-star-fill",
  size: "bs-icon-default",
  hoverEffect: true
};

describe("FormRatingWidget Component", () => {
  it("renders FormRatingWidget with selected icons and help text", () => {
    pageRenderer.render({
      body: [new FormRatingWidget(widgetProps)]
    });

    const title = screen.getByText(widgetProps.title);
    const ratingIcons = document.querySelectorAll(".rating-icon");

    expect(title).toBeInTheDocument();

    fireEvent.click(ratingIcons[0]);
    fireEvent.click(ratingIcons[1]);
    fireEvent.click(ratingIcons[2]);

    const hiddenInput = document.querySelector('[type="hidden"]')!;
    expect(hiddenInput).toHaveValue("3");
  });

  it("renders FormRatingWidget without selected icons or help text", () => {
    pageRenderer.render({
      body: [new FormRatingWidget({ ...widgetProps, helpText: "" })]
    });

    const title = screen.getByText(widgetProps.title);
    const ratingIcons = document.querySelectorAll(".rating-icon");

    expect(title).toBeInTheDocument();
    expect(ratingIcons).toHaveLength(5);

    // No selection, so the hidden input should have an empty value
    const hiddenInput = document.querySelector('[type="hidden"]')!;
    expect(hiddenInput).toHaveValue("");
  });

  it("renders correctly", () => {
    const tree = pageRenderer.getAsJSON({
      body: [new FormRatingWidget(widgetProps)]
    });
    expect(tree).toMatchSnapshot();
  });
});
