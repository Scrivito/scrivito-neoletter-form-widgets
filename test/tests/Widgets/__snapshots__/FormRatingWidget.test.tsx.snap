// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormRatingWidget Component renders correctly 1`] = `
<div>
  <div>
    <div
      className="form-rating mb-3"
    >
      <div
        className="rating-title"
      >
        <span
          className="text-super"
        >
          Test Rating Title
        </span>
        <i
          className="bi bi-question-circle"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        />
      </div>
      <i
        className="bi bi-star-fill bi-default rating-icon"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          {
            "color": "",
            "paddingRight": "0.08em",
          }
        }
      />
      <i
        className="bi bi-star-fill bi-default rating-icon"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          {
            "color": "",
            "paddingRight": "0.08em",
          }
        }
      />
      <i
        className="bi bi-star-fill bi-default rating-icon"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          {
            "color": "",
            "paddingRight": "0.08em",
          }
        }
      />
      <i
        className="bi bi-star-fill bi-default rating-icon"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          {
            "color": "",
            "paddingRight": "0.08em",
          }
        }
      />
      <i
        className="bi bi-star-fill bi-default rating-icon"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          {
            "color": "",
            "paddingRight": "0.08em",
          }
        }
      />
      <input
        className="show-in-review"
        name=""
        type="hidden"
        value=""
      />
    </div>
  </div>
</div>
`;
